lib/libxdp/xsk.c

```diff
@@ -0,0 +1,947 @@
+	prog_info.nr_map_ids = num_maps;
+	prog_info.map_ids = (__u64)(unsigned long)map_ids;
+
+	err = bpf_obj_get_info_by_fd(ctx->prog_fd, &prog_info, &prog_len);
+	if (err)
+		goto out_map_ids;
+
+	ctx->xsks_map_fd = -1;
+
+	for (i = 0; i < prog_info.nr_map_ids; i++) {
+		fd = bpf_map_get_fd_by_id(map_ids[i]);
+		if (fd < 0)
+			continue;
+
+		err = bpf_obj_get_info_by_fd(fd, &map_info, &map_len);
+		if (err) {
+			close(fd);
+			continue;
+		}
+
+		if (!strcmp(map_info.name, "xsks_map")) {
+			ctx->xsks_map_fd = fd;
+			continue;
+		}
+
+		close(fd);
+	}
+
+	err = 0;
+	if (ctx->xsks_map_fd == -1)
+		err = -ENOENT;
+
+out_map_ids:
+	free(map_ids);
+	return err;
+}
+
+static int xsk_add_map_entry(struct xsk_socket *xsk)
+{
+	struct xsk_ctx *ctx = xsk->ctx;
+
+	return bpf_map_update_elem(ctx->xsks_map_fd, &ctx->queue_id,
+				   &xsk->fd, 0);
+}
+
+static int xsk_create_xsk_struct(int ifindex, struct xsk_socket *xsk)
+{
+	char ifname[IFNAMSIZ];
+	struct xsk_ctx *ctx;
+	char *interface;
+
+	ctx = calloc(1, sizeof(*ctx));
+	if (!ctx)
+		return -ENOMEM;
+
+	interface = if_indextoname(ifindex, &ifname[0]);
+	if (!interface) {
+		free(ctx);
+		return -errno;
+	}
+
+	ctx->ifindex = ifindex;
+	memcpy(ctx->ifname, ifname, IFNAMSIZ -1);
+	ctx->ifname[IFNAMSIZ - 1] = 0;
+
+	xsk->ctx = ctx;
+
+	return 0;
+}
+
+static enum xdp_attach_mode xsk_convert_xdp_flags(__u32 xdp_flags)
+{
+	if (xdp_flags & XDP_FLAGS_REPLACE)
+		pr_warn("XDP_FLAGS_REPLACE not supported by libxdp.\n");
+
+	if (xdp_flags & XDP_FLAGS_SKB_MODE)
+		return XDP_MODE_SKB;
+	if (xdp_flags & XDP_FLAGS_DRV_MODE)
+		return XDP_MODE_NATIVE;
+	if (xdp_flags & XDP_FLAGS_HW_MODE)
+		return XDP_MODE_HW;
+
+	return XDP_MODE_UNSPEC;
+}
+
+static int xsk_lookup_program(struct xsk_socket *xsk)
+{
+	struct xdp_multiprog *multi_prog;
+	struct xdp_program *prog = NULL;
+	struct xsk_ctx *ctx = xsk->ctx;
+
+	multi_prog = xdp_multiprog__get_from_ifindex(ctx->ifindex);
+	if (!multi_prog)
+		return -1;
+
+	while ((prog = xdp_multiprog__next_prog(prog, multi_prog)))
+		if (!strcmp(xdp_program__name(prog), "xsk_def_prog"))
+			break;
+
+	if (!prog)
+		return -1;
+	return xdp_program__fd(prog);
+}
+
+static int __xsk_setup_xdp_prog(struct xsk_socket *xsk, int *xsks_map_fd)
+{
+	struct xsk_ctx *ctx = xsk->ctx;
+	int prog_fd, err;
+
+	prog_fd = xsk_lookup_program(xsk);
+
+	if (prog_fd < 0) {
+		ctx->xdp_prog = xdp_program__find_file("xsk_def_xdp_prog.o", NULL, NULL);
+		if (IS_ERR(ctx->xdp_prog))
+			return PTR_ERR(ctx->xdp_prog);
+
+		err = xsk_size_map(xsk);
+		if (err)
+			goto err_prog_load;
+
+		err = xdp_program__attach(ctx->xdp_prog, ctx->ifindex,
+					  xsk_convert_xdp_flags(xsk->config.xdp_flags), 0);
+		if (err)
+			goto err_prog_load;
+		ctx->prog_fd = xdp_program__fd(ctx->xdp_prog);
+	} else {
+		ctx->prog_fd = prog_fd;
+		if (ctx->prog_fd < 0)
+			return -errno;
+		err = xsk_lookup_bpf_map(xsk);
+		if (err) {
+			close(ctx->prog_fd);
+			return err;
+		}
+	}
+
+	if (xsk->rx) {
+		err = xsk_add_map_entry(xsk);
```

Comment:
So if no XDP program is loaded, this will attach the program before the map is populated. Isn't it better to populate the map first so we're reasonable sure everything is setup before it starts processing packets? Seeing as program attach can have side effects, we'd want to avoid doing attach-then-detach if possible?

Reply:
I can reverse the order. Actually saves some processing time during setup.